cpp/daal/src/algorithms/tsne/tsne_gradient_descent_impl.i

```diff
@@ -0,0 +1,1050 @@
+                    const DataType dy   = py - posy[n];
+                    const DataType dxy1 = dx * dx + dy * dy + EPS_PLUS_1;
+
+                    if ((n < N) or (dxy1 >= dq[depth]))
+                    {
+                        const DataType tdist_2 = mass[n] / (dxy1 * dxy1);
+                        localSum[0] += tdist_2 * dxy1;
+                        vx += dx * tdist_2;
+                        vy += dy * tdist_2;
+                    }
+                    else
+                    {
+                        pos[depth]  = pd;
+                        node[depth] = nd;
+                        depth++;
+                        pd = 0;
+                        nd = n * 4;
+                    }
+                }
+
+            } while (--depth >= 0); // done with this level
+
+            // update velocity
+            repx[i] += vx;
+            repy[i] += vy;
+        }
+    });
+    posTlsData.reduce([&](IdxType * buf) { services::internal::service_scalable_free<IdxType, cpu>(buf); });
+    nodeTlsData.reduce([&](IdxType * buf) { services::internal::service_scalable_free<IdxType, cpu>(buf); });
+    sumTlsData.reduceTo(&zNorm, 1);
+    services::internal::service_free<DataType, cpu>(dq);
+
+    return services::Status();
+}
+
+template <bool DivComp, typename IdxType, typename DataType, daal::CpuType cpu>
+services::Status attractiveKernelImpl(const DataType * val, const size_t * col, const size_t * row, const DataType * posx, const DataType * posy,
+                                      DataType * attrx, DataType * attry, DataType & zNorm, DataType & divergence, const IdxType nNodes,
+                                      const IdxType N, const IdxType nnz, const IdxType nElements, const DataType exaggeration)
+{
+    DAAL_CHECK_MALLOC(val);
+    DAAL_CHECK_MALLOC(col);
+    DAAL_CHECK_MALLOC(row);
+    DAAL_CHECK_MALLOC(posx);
+    DAAL_CHECK_MALLOC(posy);
+    DAAL_CHECK_MALLOC(attrx);
+    DAAL_CHECK_MALLOC(attry);
+
+    //initialize arrays
+    services::internal::service_memset<DataType, cpu>(attrx, 0., N);
+    services::internal::service_memset<DataType, cpu>(attry, 0., N);
+
+    const DataType multiplier = exaggeration * (DataType)zNorm;
+    divergence                = 0.;
+
+    daal::StaticTlsSum<DataType, cpu> divTlsData(1);
+    daal::static_tls<DataType *> logTlsData([=]() { return services::internal::service_scalable_calloc<DataType, cpu>(nElements); });
+
+    const size_t nThreads    = logTlsData.nthreads();
+    const size_t sizeOfBlock = services::internal::min<cpu, size_t>(256, N / nThreads + 1);
+    const size_t nBlocks     = (size_t)N / sizeOfBlock + !!((size_t)N % sizeOfBlock);
+
+    daal::static_threader_for(nBlocks, [&](size_t iBlock, size_t tid) {
+        const size_t iStart = iBlock * sizeOfBlock;
+        const size_t iEnd   = services::internal::min<cpu, size_t>((size_t)N, iStart + sizeOfBlock);
+        DataType * log      = logTlsData.local(tid);
```

Comment:
log - standart name in C and C++ libraries. These can happen conflict. Use other name
```suggestion
        DataType * logLocal      = logTlsData.local(tid);
```